### ISSUE TITLE
Frida gadget 12.8.7 on Win32: instrumented process remains alive in a sort of "zombie" state after quitting it

### ISSUE THREAD

**Issue by Palacee-hun** [frida/frida#1159]
I have written an instrumentation with Frida for an app. It turned out to be very useful and I have been using it a lot, so I thought I'd try to package it together with the app with the help of the autonomous script mode of Frida gadget.

So I copied my instrumentation js script to the directory of the app and I also copied frida gadget dll there renamed as helper.dll. I also created helper.config, setting interactions to "script" and on_change to "reload" in it and of course the path of my script.

Then as a first try I injected frida gadget dll into the running app with two tools: inject.exe from Comrade's PETOOLS and procwin.exe written by Japheth. My instrumentation worked flawlessly with both tools and the app quit all right. Then I took the plunge and patched the app executable with patch.exe also from Comrade's PETOOLS so that it would load frida gadget dll first upon starting before doing anything else. This was also successful: the instrumented app did what it was supposed to do and quit normally.

After all this I was surprised to still see instances of the instrumented process listed in my process explorer, although they had quit normally, their windows had been closed and I saw no visible traces of them running.

So I investigated those "zombie" processes and I saw that they unloaded some of their dependent dlls, but not all. And helper.dll (frida gadget dll) was still loaded in all cases together with its dependency dlls. It was always possible to kill the zombie processes easily. So somehow frida gadget dll didn't detach from the process when it quit.

I have tried all this with rpc.exports and init/dispose in my js script and also without them: this made absolutely no difference and the results were the exact same. I experimented with other processes and instrumentations and this zombie process thing was 100 % reproducible when doing it with frida gadget dll, but it has never eve happened when using the python frida toolset. I tried to temporarily disable my antivirus too, but it also did not matter.

My system is Windows 7 Ultimate 32-bit.

**Comment by Palacee-hun** [comment]
Anyway my instrumentation was quite simple: one Interceptor.attach, two Nativefunctions and a minimal branching logic in javascript.

**Comment by Palacee-hun** [comment]
I've experimented further and found out that neither "teardown" nor "runtime" config key mattered in this respect: the issue is always reproducible even with full teardown set in the config and also with jit as runtime. There is always one thread encountered by the process explorer in the "zombie" process.

**Comment by Palacee-hun** [comment]
After doing a lot of testing with more recent versions of Frida gadget (e.g. 14.x) and various versions of Windows, I have edited the issue title.
It persists on the latest version (14.2.2 now) also. Windows version also doesn't seem to matter as the issue is reproducible on the latest Windows 10 and presents identically. 64-bit Windows gadgets are also affected. I have also discovered that The contents of the instrumentation script do not matter in this respect, the issue is always reproducible even with an empty script file.
The process which has the gadget dll loaded simply does not terminate. I have investigated many such zombie processes with Sysinternals Process Explorer and it seems that the gadget dll still holds the handles to all the threads of the instrumented process even though they have already stopped and does not release those handles.
I would gladly try my hand at tracking down the cause in the sources if someone were kind enough to point me at where the Windows-specific termination code lies in the giant source "forest".

**Comment by cesaryuan** [comment]
Sorry to bother you after so long, I discounted the same problem with you. Did you find a method to solve it?

My situation is this:

1. Use DLL hijacking technology to make a version.dll and place it under the target application directory, so that my forged version.dll will be loaded when the target application starts. Then I added the `LoadLibraryW(L"frida-gadget.dll" to version.dll );` command to start the gadget. The interaction mode I used is `Script` and it worked fine.
2. However, when I exit the target application, the target application still stays in the process list.

My system and Frida information: Windows 10 22H2, Frida gadget 16.0.18.

I would really appreciate it if someone could take another look at this problem because this application mode (automatic hooking by hijacking the gadget into the target application) is very convenient.

**Comment by Palacee-hun** [comment]
No, unfortunately I haven't found any solution or workaround for this
scenario in the framework of Frida. Your letter is another proof that
the scenario I made the issue about is independent of Windows version
or frida-gadget version. This issue is most probably around for many
years now, and probably much much older than Frida version 14.
For the specific situation that brought up the issue for me I simply
wrote a very tiny instrumentation dll by hand. This was quite easy to
do in that specific case as the instrumentation needed was extremely
simple. My instrumentation dll just gets r/w access to the executable
section with VirtualProtect windows call, then it redirects control
flow to my little code by inserting a jmp at the right point and
finally restores r/x protection to the executable section. In this
specific case this is totally satisfactory and loads instantaneously
(I make the target app load my dll with an exe patcher utility). This
is in place ever since and works perfectly.
If I want to use the autonomous script mode of Frida gadget for a much
more complex case (I agree, it is very neat and convenient), then I
still have to put up with the zombie processes. I kill them once in a
while withthe procwin utility by Japheth (found at japheth.de). I
don't bother killing them right after they are created (i.e. after
each run of the gadgeted process), because they are harmless and hog
no noticable system resources.
I made many attempts to try to discover what goes on behind the scenes
here, but Frida code has many many layers and extremely complex and
diverse, and I couldn't determine what it is that Frida gadget does at
unloading and a normal injection doesn't (normal injection never ever
produces zombie processes). I suspected some form of channel was still
open and that would keep the process in the task list, but I couldn't
find any, no matter how many port scanners or resource monitors I
tried.
As the main author of Frida hasn't reacted to this issue, I am afraid
he is clueless, and that is a bad thing as he seems to be the only one
who sees this very huge codebase in its entirety with its odds and
ends. Thhe only real hope for this seems to be the injector rewrite
that started in 16.0.12 and planned to come also to Windows
eventually. It maynot be in the near future though as most work is
being done for Darwin, Linux and Android. This is quite evident from
the changelogs.


2023-04-23 9:37 GMT+02:00, Cesaryuan ***@***.***>:
